nexus/reconfigurator/planning/src/blueprint_builder/builder/disks_editor.rs

```diff
@@ -0,0 +1,178 @@
+
+use nexus_types::deployment::BlueprintPhysicalDiskConfig;
+use nexus_types::deployment::BlueprintPhysicalDisksConfig;
+use omicron_common::api::external::Generation;
+use omicron_uuid_kinds::GenericUuid;
+use omicron_uuid_kinds::PhysicalDiskUuid;
+use omicron_uuid_kinds::SledUuid;
+use std::collections::BTreeMap;
+use std::collections::BTreeSet;
+
+/// Helper for working with sets of disks on each sled
+///
+/// Tracking the set of disks is slightly non-trivial because we need to
+/// bump the per-sled generation number iff the disks are changed.  So
+/// we need to keep track of whether we've changed the disks relative
+/// to the parent blueprint.
+#[derive(Debug)]
+pub(super) struct BlueprintDisksEditor {
+    current: BTreeMap<SledUuid, DisksConfig>,
+    changed: BTreeSet<SledUuid>,
+}
+
+impl BlueprintDisksEditor {
+    pub fn new(
+        current: BTreeMap<SledUuid, BlueprintPhysicalDisksConfig>,
+    ) -> Self {
+        let current = current
+            .into_iter()
+            .map(|(sled_id, config)| (sled_id, config.into()))
+            .collect();
+        Self { current, changed: BTreeSet::new() }
+    }
+
+    /// Get a helper to edit the disks of a specific sled.
+    ///
+    /// If any changes are made via the returned editor, the sled will be
+    /// recorded as needing a generation bump in its disk config when the editor
+    /// is dropped.
+    pub fn sled_disks_editor(
+        &mut self,
+        sled_id: SledUuid,
+    ) -> SledDisksEditor<'_> {
+        let config =
+            self.current.entry(sled_id).or_insert_with(DisksConfig::empty);
+        SledDisksEditor::new(sled_id, config, &mut self.changed)
+    }
+
+    #[cfg(test)]
+    pub fn current_sled_disks(
+        &self,
+        sled_id: &SledUuid,
+    ) -> Option<&BTreeMap<PhysicalDiskUuid, BlueprintPhysicalDiskConfig>> {
+        let config = self.current.get(sled_id)?;
+        Some(&config.disks)
+    }
+
+    pub fn build(
+        mut self,
+        sled_ids: impl Iterator<Item = SledUuid>,
+    ) -> BTreeMap<SledUuid, BlueprintPhysicalDisksConfig> {
+        sled_ids
+            .map(|sled_id| {
+                let config = match self.current.remove(&sled_id) {
+                    Some(mut config) => {
+                        // Bump generation number for any sled whose DisksConfig
+                        // changed
+                        if self.changed.contains(&sled_id) {
+                            config.generation = config.generation.next()
+                        }
+                        config.into()
+                    }
+                    None => DisksConfig::empty().into(),
+                };
+                (sled_id, config)
+            })
+            .collect()
+    }
+}
+
+#[derive(Debug)]
+pub(super) struct SledDisksEditor<'a> {
+    config: &'a mut DisksConfig,
+    changed: bool,
+    sled_id: SledUuid,
+    parent_changed_set: &'a mut BTreeSet<SledUuid>,
+}
+
+impl Drop for SledDisksEditor<'_> {
```

Comment:
Neat, so the mutable borrows here require that `SledDisksEditor` is dropped before `BlueprintDisksEditor::build` gets called, right?

Reply:
Correct.

You could always leak or `mem::forget()` a `SledDisksEditor`, but at that point you get what you deserve, I think? The alternative would be for `sled_disks_editor` to take an `FnOnce(&mut SledDisksEditor)`, but that's much more awkward from a caller's point of view. It seems fine to say "if you leak a `SledDisksEditor` your edits won't be fully reflected".